File: android-sdk-ui/src/com/appboy/ui/inappmessage/listeners/TouchAwareSwipeDismissTouchListener.java
package com.appboy.ui.inappmessage.listeners;

import android.view.MotionEvent;
import android.view.View;

/**
 * Adds touch events to the SwipeDismissTouchListener.
 *
 * Note: The SwipeDismissTouchListener requires API level 12 or later due to use of {@link
 * android.view.ViewPropertyAnimator}.</p>
 */
public class TouchAwareSwipeDismissTouchListener extends SwipeDismissTouchListener {
  private ITouchListener mTouchListener;

  public interface ITouchListener {
    void onTouchStartedOrContinued();

    void onTouchEnded();
  }

  public TouchAwareSwipeDismissTouchListener(View view, Object token, DismissCallbacks callbacks) {
    super(view, token, callbacks);
  }

  public void setTouchListener(ITouchListener touchListener) {
    mTouchListener = touchListener;
  }

  @Override
  public boolean onTouch(View view, MotionEvent motionEvent) {
    switch (motionEvent.getAction()) {
      case MotionEvent.ACTION_DOWN:
        if (mTouchListener != null) {
          mTouchListener.onTouchStartedOrContinued();
        }
        break;
      case MotionEvent.ACTION_UP:
      case MotionEvent.ACTION_CANCEL:
        if (mTouchListener != null) {
          mTouchListener.onTouchEnded();
        }
        break;
      default:
        break;
    }
    return super.onTouch(view, motionEvent);
  }
}

File: android-sdk-ui/src/main/java/com/braze/ui/inappmessage/listeners/TouchAwareSwipeDismissTouchListener.kt
package com.braze.ui.inappmessage.listeners

import android.view.MotionEvent
import android.view.View

/**
 * Adds touch events to the SwipeDismissTouchListener.
 */
class TouchAwareSwipeDismissTouchListener(view: View, token: Any?, callbacks: DismissCallbacks?) :
    SwipeDismissTouchListener(view, token, callbacks) {
    private var touchListener: ITouchListener? = null

    interface ITouchListener {
        fun onTouchStartedOrContinued()
        fun onTouchEnded()
    }

    constructor(view: View, callbacks: DismissCallbacks?) : this(view, null, callbacks)

    fun setTouchListener(newTouchListener: ITouchListener) {
        touchListener = newTouchListener
    }

    override fun onTouch(view: View, motionEvent: MotionEvent): Boolean {
        when (motionEvent.action) {
            MotionEvent.ACTION_DOWN -> touchListener?.onTouchStartedOrContinued()
            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> touchListener?.onTouchEnded()
            else -> {}
        }
        return super.onTouch(view, motionEvent)
    }
}
